lib/pengine/pe_actions.c

```diff
@@ -446,66 +446,87 @@ validate_on_fail(const pe_resource_t *rsc, const char *action_name,
 {
     const char *name = NULL;
     const char *role = NULL;
-    const char *on_fail = NULL;
     const char *interval_spec = NULL;
     const char *value = g_hash_table_lookup(meta, XML_OP_ATTR_ON_FAIL);
 
-    if (pcmk__str_eq(action_name, PCMK_ACTION_STOP, pcmk__str_casei)
+    // Stop actions can only use certain on-fail values
+    if (pcmk__str_eq(action_name, PCMK_ACTION_STOP, pcmk__str_none)
         && !valid_stop_on_fail(value)) {
 
         pcmk__config_err("Resetting '" XML_OP_ATTR_ON_FAIL "' for %s stop "
                          "action to default value because '%s' is not "
                          "allowed for stop", rsc->id, value);
         return NULL;
+    }
 
-    } else if (pcmk__str_eq(action_name, PCMK_ACTION_DEMOTE, pcmk__str_casei)
-               && (value == NULL)) {
-        // demote on_fail defaults to monitor value for promoted role if present
-        xmlNode *operation = NULL;
-
-        CRM_CHECK(rsc != NULL, return NULL);
+    /* Demote actions default on-fail to the on-fail value for the first
+     * recurring monitor for the promoted role (if any).
+     */
+    if (pcmk__str_eq(action_name, PCMK_ACTION_DEMOTE, pcmk__str_none)
+        && (value == NULL)) {
 
-        for (operation = pcmk__xe_first_child(rsc->ops_xml);
-             (operation != NULL) && (value == NULL);
-             operation = pcmk__xe_next(operation)) {
+        /* @TODO This does not consider promote options set in a meta-attribute
+         * block (which may have rules that need to be evaluated) rather than
+         * XML properties.
+         */
+        for (xmlNode *operation = first_named_child(rsc->ops_xml, XML_ATTR_OP);
+             (operation != NULL); operation = crm_next_same_xml(operation)) {
```

Comment:
No parentheses needed around the loop condition now

Reply:
thanks